lib/ramble/ramble/cmd/deployment.py

```diff
@@ -0,0 +1,178 @@
+    'publish',
+    'fetch',
+]
+
+
+def deployment_publish_setup_parser(subparser):
+    """Publish a workspace deployment"""
+    subparser.add_argument(
+        '--tar-archive', '-t', action='store_true',
+        dest='tar_archive',
+        help='create a tar.gz of the deployment directory.')
+
+    subparser.add_argument(
+        '--deployment-name', '-d', dest='deployment_name',
+        default=None,
+        help='Name for deployment. Uses workspace name if not set.')
+
+    subparser.add_argument(
+        '--upload-url', '-u', dest='upload_url',
+        default=None,
+        help='URL to upload deployment into. Upload tar if `-t` is specified..')
+
+    arguments.add_common_arguments(subparser, ['phases', 'include_phase_dependencies',
+                                               'where', 'exclude_where', 'filter_tags'])
+
+
+def deployment_publish(args):
+    current_pipeline = ramble.pipeline.pipelines.publishdeployment
+    ws = ramble.cmd.require_active_workspace(cmd_name='deployment publish')
+
+    filters = ramble.filters.Filters(
+        phase_filters=args.phases,
+        include_where_filters=args.where,
+        exclude_where_filters=args.exclude_where,
+        tags=args.filter_tags
+    )
+
+    pipeline_cls = ramble.pipeline.pipeline_class(current_pipeline)
+
+    pipeline = pipeline_cls(ws,
+                            filters,
+                            create_tar=args.tar_archive,
+                            upload_url=args.upload_url,
+                            deployment_name=args.deployment_name)
+
+    with ws.write_transaction():
+        deployment_run_pipeline(args, pipeline)
+
+
+def deployment_fetch_setup_parser(subparser):
+    """Fetch a workspace deployment into current workspace"""
+    subparser.add_argument(
+        '--deployment-path', '-p', dest='deployment_path',
+        help='Path to deployment that should be fetched')
+
+
+def deployment_fetch(args):
+    def fetch_file(src, dest):
+        fetcher = ramble.fetch_strategy.URLFetchStrategy(url=src)
+        stage_dir = os.path.dirname(dest)
+        fs.mkdirp(stage_dir)
+        with ramble.stage.InputStage(fetcher, path=stage_dir,
+                                     name=os.path.basename(src)) as stage:
+            stage.fetch()
+
+    ws = ramble.cmd.require_active_workspace(cmd_name='deployment fetch')
+
+    with ws.write_transaction():
+        # Fetch deployment index first:
+        publish_cls = ramble.pipeline.PublishDeploymentPipeline
+
+        remote_index_path = args.deployment_path + \
+            '/' + ramble.pipeline.PublishDeploymentPipeline.index_filename
+        local_index_path = \
+            os.path.join(ws.root, publish_cls.index_filename)
+
+        fetch_file(remote_index_path, local_index_path)
+
+        with open(local_index_path, 'r') as f:
+            index_data = sjson.load(f)
+
+        for file in index_data[publish_cls.index_namespace]:
+            src = args.deployment_path + '/' + file
```

Comment:
We use this pattern here, but we should probably rely on something like `os.join` since paths with too many `/` break in gcs buckets

Reply:
My main concern was in URLs that might not want to use the OS separator (i.e. windows). But, this would also break local deployments at the OS level if you were publishing to a different directory.

Do you have any thoughts on how to manage this aspect of it?

Reply:
So more like `urllib`'s `urljoin`?

Reply:
Yes! But I also think it might be trickier than that... I wonder what urllib does if the root is `file://` Maybe I'll play with it some and see if that would be a better replacement than just injecting `/`.